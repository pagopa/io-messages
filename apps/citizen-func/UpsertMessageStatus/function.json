{
  "bindings": [
    {
      "authLevel": "function",
      "type": "httpTrigger",
      "direction": "in",
      "name": "req",
      "route": "v1/messages/{fiscal_code}/{id}/message-status",
      "methods": [
        "put"
      ]
    },
    {
      "type": "http",
      "direction": "out",
      "name": "res"
    }
  ],
  "scriptFile": "../dist/UpsertMessageStatus/index.js"
}