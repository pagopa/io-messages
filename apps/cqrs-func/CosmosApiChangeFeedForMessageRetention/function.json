{
  "bindings": [
    {
      "type": "cosmosDBTrigger",
      "name": "messageStatusDocuments",
      "direction": "in",
      "leaseContainerName": "leaseMessageStatusForMessageRetention",
      "leaseContainerPrefix": "%LEASE_COLLECTION_PREFIX%_",
      "connection": "COSMOSDB_CONNECTION_STRING",
      "databaseName": "%COSMOSDB_NAME%",
      "containerName": "message-status",
      "createLeaseContainerIfNotExists": true,
      "startFromBeginning": true
    }
  ],
  "retry": {
    "strategy": "exponentialBackoff",
    "maxRetryCount": 10,
    "minimumInterval": "00:00:05",
    "maximumInterval": "00:01:00"
  },
  "scriptFile": "../dist/CosmosApiChangeFeedForMessageRetention/index.js"
}
