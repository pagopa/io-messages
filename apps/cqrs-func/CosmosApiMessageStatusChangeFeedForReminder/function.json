{
  "bindings": [
    {
      "authLevel": "function",
      "type": "cosmosDBTrigger",
      "name": "cosmosApiMessageStatusTrigger",
      "direction": "in",
      "connection": "COSMOSDB",
      "databaseName": "%COSMOSDB_NAME%",
      "containerName": "message-status",
      "leaseContainerName": "cqrs-leases",
      "leaseContainerPrefix": "CosmosApiMessageStatusChangeFeedForReminder",
      "createLeaseContainerIfNotExists": true,
      "startFromBeginning": false
    }
  ],
  "retry": {
    "strategy": "fixedDelay",
    "maxRetryCount": -1,
    "delayInterval": "00:00:10"
  },
  "scriptFile": "../dist/CosmosApiMessageStatusChangeFeedForReminder/index.js"
}
