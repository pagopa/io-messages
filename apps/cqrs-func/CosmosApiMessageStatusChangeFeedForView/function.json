{
  "bindings": [
    {
      "authLevel": "function",
      "type": "cosmosDBTrigger",
      "name": "cosmosApiMessageStatusTrigger",
      "direction": "in",
      "connection": "COSMOSDB_CONNECTION_STRING",
      "databaseName": "%COSMOSDB_NAME%",
      "containerName": "message-status",
      "leaseContainerName": "cqrs-leases",
      "leaseContainerPrefix": "CosmosApiMessageStatusChangeFeedForView",
      "createLeaseContainerIfNotExists": true,
      "startFromBeginning": true
    }
  ],
  "retry": {
    "strategy": "fixedDelay",
    "maxRetryCount": -1,
    "delayInterval": "00:00:10"
  },
  "scriptFile": "../dist/CosmosApiMessageStatusChangeFeedForView/index.js"
}