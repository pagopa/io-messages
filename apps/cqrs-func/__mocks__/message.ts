import * as TE from "fp-ts/lib/TaskEither";

import { FeatureLevelTypeEnum } from "@pagopa/io-functions-commons/dist/generated/definitions/FeatureLevelType";
import { MessageContent } from "@pagopa/io-functions-commons/dist/generated/definitions/MessageContent";
import { NotRejectedMessageStatusValueEnum } from "@pagopa/io-functions-commons/dist/generated/definitions/NotRejectedMessageStatusValue";
import { ServiceId } from "@pagopa/io-functions-commons/dist/generated/definitions/ServiceId";
import { TimeToLiveSeconds } from "@pagopa/io-functions-commons/dist/generated/definitions/TimeToLiveSeconds";
import {
  Components,
  MessageView,
  Status
} from "@pagopa/io-functions-commons/dist/src/models/message_view";
import { NonNegativeInteger } from "@pagopa/ts-commons/lib/numbers";
import { FiscalCode, NonEmptyString } from "@pagopa/ts-commons/lib/strings";
import * as E from "fp-ts/lib/Either";
import {
  MessageModel,
  RetrievedMessageWithoutContent
} from "@pagopa/io-functions-commons/dist/src/models/message";
import { MessageStatusModel } from "@pagopa/io-functions-commons/dist/src/models/message_status";
import { CosmosErrors } from "@pagopa/io-functions-commons/dist/src/utils/cosmosdb_model";
import { vi } from "vitest";
import { RetrievedMessageStatusWithFiscalCode } from "../utils/message_status";

export const now = new Date();

export const aFiscalCode = "FRLFRC74E04B157I" as FiscalCode;
export const aMessageId = "A_MESSAGE_ID" as NonEmptyString;

export const cosmosMetadata = {
  _etag: "_etag",
  _rid: "_rid",
  _self: "_self",
  _ts: 1
};

export const aRetrievedMessageWithoutContent: RetrievedMessageWithoutContent = {
  ...cosmosMetadata,
  featureLevelType: FeatureLevelTypeEnum.STANDARD,
  isPending: false,
  fiscalCode: aFiscalCode,
  id: aMessageId,
  indexedId: "A_MESSAGE_ID" as NonEmptyString,
  senderServiceId: "agid" as ServiceId,
  senderUserId: "u123" as NonEmptyString,
  timeToLiveSeconds: 3600 as TimeToLiveSeconds,
  createdAt: new Date(),
  kind: "IRetrievedMessageWithoutContent"
};

export const aMessageBodyMarkdown = "test".repeat(80);
export const aMessageContent = E.getOrElseW(() => {
  throw new Error();
})(
  MessageContent.decode({
    markdown: aMessageBodyMarkdown,
    subject: "test".repeat(10)
  })
);

export const aMessageContentWithDueDate = E.getOrElseW(() => {
  throw new Error();
})(
  MessageContent.decode({
    markdown: aMessageBodyMarkdown,
    subject: "test".repeat(10),
    due_date: now
  })
);
export const aMessageStatus: RetrievedMessageStatusWithFiscalCode = {
  ...cosmosMetadata,
  messageId: aMessageId,
  id: `${aMessageId}-0` as NonEmptyString,
  status: NotRejectedMessageStatusValueEnum.PROCESSED,
  version: 0 as NonNegativeInteger,
  updatedAt: new Date(),
  fiscalCode: aFiscalCode,
  isRead: false,
  isArchived: false,
  kind: "IRetrievedMessageStatus"
};

export const aComponents: Components = {
  attachments: { has: false },
  euCovidCert: { has: false },
  legalData: { has: false },
  payment: { has: false },
  thirdParty: { has: false }
};

export const aStatus: Status = {
  archived: false,
  processing: NotRejectedMessageStatusValueEnum.PROCESSED,
  read: false
};

export const aMessageView: MessageView = {
  components: aComponents,
  createdAt: now,
  fiscalCode: aFiscalCode,
  id: aMessageId,
  messageTitle: "a-msg-title" as NonEmptyString,
  senderServiceId: "a-service-id" as ServiceId,
  status: aStatus,
  version: 0 as NonNegativeInteger
};

export const mockPatch = vi.fn(
  (): TE.TaskEither<CosmosErrors, RetrievedMessageWithoutContent> =>
    TE.of(aRetrievedMessageWithoutContent)
);

export const mockMessageModel: MessageModel = ({
  patch: mockPatch
} as any) as MessageModel;

export const mockUpdateTTLForAllVersions = vi.fn(
  (): TE.TaskEither<CosmosErrors, number> => TE.of(1)
);

export const mockMessageStatusModel: MessageStatusModel = ({
  updateTTLForAllVersions: mockUpdateTTLForAllVersions
} as unknown) as MessageStatusModel;
