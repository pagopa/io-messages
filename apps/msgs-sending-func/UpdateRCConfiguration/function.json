{
  "bindings": [
    {
      "authLevel": "function",
      "type": "httpTrigger",
      "direction": "in",
      "name": "req",
      "route": "v1/remote-contents/configurations/{configurationId}",
      "methods": [
        "put"
      ]
    },
    {
      "type": "http",
      "direction": "out",
      "name": "res"
    }
  ],
  "scriptFile": "../dist/UpdateRCConfiguration/index.js"
}
