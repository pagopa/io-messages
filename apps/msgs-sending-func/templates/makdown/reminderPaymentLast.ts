/* eslint-disable sort-keys */
import { IPrintersForTemplate } from "../printer";
import {
  OPEN_THE_APP_TO_PAY_IT,
  YOU_HAVE_A_NOTICE_EXPIRING_TOMORROW
} from "./constant";

export const reminderPaymentLastPrinter: IPrintersForTemplate = {
  silentPushPrinter: _ne => ({
    title: YOU_HAVE_A_NOTICE_EXPIRING_TOMORROW,
    body: OPEN_THE_APP_TO_PAY_IT
  }),

  verbosePushPrinter: ne => ({
    title: YOU_HAVE_A_NOTICE_EXPIRING_TOMORROW,
    body: `Entra nell’app e paga l’avviso emesso da ${ne.organizationName}`
  })
};
