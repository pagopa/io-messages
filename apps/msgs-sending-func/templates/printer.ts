/* eslint-disable @typescript-eslint/naming-convention */
import * as t from "io-ts";
import { NonEmptyString } from "@pagopa/ts-commons/lib/strings";
import {
  NotificationType,
  NotificationTypeEnum
} from "../generated/definitions/NotificationType";
import { messagePrinter } from "./makdown/message";
import { reminderReadPrinter } from "./makdown/reminderRead";
import { reminderPaymentPrinter } from "./makdown/reminderPayment";
import { reminderPaymentLastPrinter } from "./makdown/reminderPaymentLast";

export const NotificationEntry = t.interface({
  organizationName: NonEmptyString,
  serviceName: NonEmptyString,
  title: NonEmptyString
});
export type NotificationEntry = t.TypeOf<typeof NotificationEntry>;

export interface NotificationPrinter {
  readonly title: string;
  readonly body: string;
}

export interface IPrintersForTemplate {
  readonly silentPushPrinter: (v: NotificationEntry) => NotificationPrinter;
  readonly verbosePushPrinter: (v: NotificationEntry) => NotificationPrinter;
}

const printersConfigurations: {
  [key in NotificationType]: IPrintersForTemplate;
} = {
  [NotificationTypeEnum.MESSAGE]: messagePrinter,
  [NotificationTypeEnum.REMINDER_READ]: reminderReadPrinter,
  [NotificationTypeEnum.REMINDER_PAYMENT]: reminderPaymentPrinter,
  [NotificationTypeEnum.REMINDER_PAYMENT_LAST]: reminderPaymentLastPrinter
};

export const getPrinterForTemplate = (
  notificationType: NotificationType
): IPrintersForTemplate => printersConfigurations[notificationType];
