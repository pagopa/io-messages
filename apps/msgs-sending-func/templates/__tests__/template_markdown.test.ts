import { getOrElseW } from "fp-ts/lib/Either";
import { pipe } from "fp-ts/lib/function";

import { NotificationEntry } from "../printer";

import { messagePrinter } from "../makdown/message";
import { reminderReadPrinter } from "../makdown/reminderRead";
import { reminderPaymentPrinter } from "../makdown/reminderPayment";
import { reminderPaymentLastPrinter } from "../makdown/reminderPaymentLast";

const notificationMessage = {
  organizationName: "TEST ORG",
  serviceName: "TEST SERVICE",
  title: "TEST OBJECT"
};

describe.each([
  { test: "MESSAGE", printer: messagePrinter },
  { test: "REMINDER_READ", printer: reminderReadPrinter },
  { test: "REMINDER_PAYMENT", printer: reminderPaymentPrinter },
  { test: "REMINDER_PAYMENT_LAST", printer: reminderPaymentLastPrinter }
])(`Printer - Template - %s`, ({ printer }) => {
  it("should print a silent notification", () => {
    const notification = pipe(
      NotificationEntry.decode(notificationMessage),
      getOrElseW(_ => {
        throw "Error decoding object";
      })
    );

    const result = printer.silentPushPrinter(notification);
    expect(result).toMatchSnapshot();
  });

  it("should print a verbose notification", () => {
    const notification = pipe(
      NotificationEntry.decode(notificationMessage),
      getOrElseW(_ => {
        throw "Error decoding object";
      })
    );

    const result = printer.verbosePushPrinter(notification);
    expect(result).toMatchSnapshot();
  });
});
