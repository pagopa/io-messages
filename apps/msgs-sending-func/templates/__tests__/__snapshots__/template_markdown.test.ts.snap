// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Printer - Template - { test: 'MESSAGE', printer: [Object] } should print a silent notification 1`] = `
Object {
  "body": "Entra nell'app per leggerlo",
  "title": "Hai un nuovo messaggio",
}
`;

exports[`Printer - Template - { test: 'MESSAGE', printer: [Object] } should print a verbose notification 1`] = `
Object {
  "body": "TEST OBJECT",
  "title": "TEST ORG",
}
`;

exports[`Printer - Template - { test: 'REMINDER_PAYMENT', printer: [Object] } should print a silent notification 1`] = `
Object {
  "body": "Entra nell'app per pagarlo",
  "title": "Hai un avviso da pagare",
}
`;

exports[`Printer - Template - { test: 'REMINDER_PAYMENT', printer: [Object] } should print a verbose notification 1`] = `
Object {
  "body": "Entra nell’app e paga l’avviso emesso da TEST ORG",
  "title": "Hai un avviso da pagare",
}
`;

exports[`Printer - Template - { test: 'REMINDER_PAYMENT_LAST', printer: [Object] } should print a silent notification 1`] = `
Object {
  "body": "Entra nell'app per pagarlo",
  "title": "Hai un avviso che scade domani",
}
`;

exports[`Printer - Template - { test: 'REMINDER_PAYMENT_LAST', printer: [Object] } should print a verbose notification 1`] = `
Object {
  "body": "Entra nell’app e paga l’avviso emesso da TEST ORG",
  "title": "Hai un avviso che scade domani",
}
`;

exports[`Printer - Template - { test: 'REMINDER_READ', printer: [Object] } should print a silent notification 1`] = `
Object {
  "body": "Entra nell'app per leggerlo",
  "title": "Hai un messaggio non letto",
}
`;

exports[`Printer - Template - { test: 'REMINDER_READ', printer: [Object] } should print a verbose notification 1`] = `
Object {
  "body": "TEST OBJECT",
  "title": "Leggi il messaggio da TEST ORG",
}
`;
